### PR TITLE
Offscreen add attach

### PR DESCRIPTION
**Pull request by sammy-SC**
Initial implementation of `Offscreen.attach`.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js [hunk 1]
```diff
@@ -2421,17 +2425,48 @@ export function detachOffscreenInstance(instance: OffscreenInstance): void {
...
+      node = node.sibling;
+    }
+  };
+
   const executionContext = getExecutionContext();
   if ((executionContext & (RenderContext | CommitContext)) !== NoContext) {
-    scheduleMicrotask(() => {
-      instance._visibility |= OffscreenDetached;
-      disappearLayoutEffects(currentOffscreenFiber);
-      disconnectPassiveEffect(currentOffscreenFiber);
```
Comment (sammy-SC) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:2429]: This is to avoid unmounting refs on offscreen in detach. Instead I unmount effects on all children.

File: packages/react-reconciler/src/__tests__/ReactOffscreen-test.js [hunk 2]
```diff
@@ -1771,5 +1833,252 @@ describe('ReactOffscreen', () => {
...
+    expect(Scheduler).toHaveYielded(['detach child']);
+
+    await act(async () => {
+      const instance = offscreen.current;
+      // Attach then immediately detach.
+      instance.attach();
+      instance.detach();
+    });
+
+    expect(Scheduler).toHaveYielded(['attach child', 'detach child']);
```
Comment (sammy-SC) [inline packages/react-reconciler/src/__tests__/ReactOffscreen-test.js:2046]: attach is synchronous, that's why child is attached and then immediately detached.

Comment (acdlite) [reply packages/react-reconciler/src/__tests__/ReactOffscreen-test.js:2046]: It should still be batched, though. That's how a discrete update works, for example. Or a setState inside a layout effect.

The model should be that it has the same behavior as a setState. Any clever tricks we do on top of that are only optimizations — like skipping the render phase if we can prove it's not necessary.

File: packages/react-reconciler/src/ReactFiberCommitWork.old.js [hunk 3]
```diff
@@ -2414,24 +2416,54 @@ function getRetryCache(finishedWork) {
...
-  } else {
+  if ((instance._visibility & OffscreenDetached) !== NoFlags) {
+    // The instance is already detached, this is a noop.
+    return;
+  }
+
+  instance._pendingVisibility |= OffscreenDetached;
+
+  // Detaching needs to be postoned in case attach is called before next update.
+  scheduleMicrotask(() => {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberCommitWork.old.js:2434]: Since all this does now is schedule an update, you don't need the `scheduleMicrotask` call

File: packages/react-reconciler/src/ReactFiberCommitWork.old.js [hunk 4]
```diff
@@ -2414,24 +2416,54 @@ function getRetryCache(finishedWork) {
...
+    if ((instance._pendingVisibility & OffscreenDetached) === NoFlags) {
+      // Attach was called. Offscreen does not need to be detached.
+      return;
+    }
+
     instance._visibility |= OffscreenDetached;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberCommitWork.old.js:2440]: This doesn't look right to me. `_pendingVisibility` is represents the queued state. `_visibility` represents the current state. So it shouldn't be set until the commit phase.

File: packages/react-reconciler/src/ReactFiberCommitWork.old.js (hunk 3 above)
Comment (sammy-SC) [reply packages/react-reconciler/src/ReactFiberCommitWork.old.js]: I was using scheduleMicrotask to postpone detachment of Offscreen if `detach` is called in an effect or in render phase. I'm not sure how to do that with `scheduleUpdateOnFiber`.

File: packages/react-reconciler/src/ReactFiberCommitWork.old.js (hunk 4 above)
Comment (sammy-SC) [reply packages/react-reconciler/src/ReactFiberCommitWork.old.js]: I changed it so `attachOffscreenInstance` and `detachOffscreenInstance` only read/write to `_pendingVisibility` and `_pendingVisibility` is synced with `_visibility` in `beginWork`.

File: packages/react-reconciler/src/__tests__/ReactOffscreen-test.js [hunk 5]
```diff
@@ -1620,45 +1663,50 @@ describe('ReactOffscreen', () => {
...
+      // Offscreen is attached and gets detached inside useLayoutEffect.
+      // State updates from offscreen are **defered**.
       await act(async () => {
         updateChildState(1);
         updateHighPriorityComponentState(1);
         expect(Scheduler).toFlushUntilNextPaint([
           'HighPriorityComponent 1',
           'Child 1',
           'HighPriorityComponent 2',
-          'Child 2',
```
Comment (sammy-SC) [inline packages/react-reconciler/src/__tests__/ReactOffscreen-test.js:1632]: @acdlite  This PR changes how `detach` works. Detach called in an effect or in render function will be executed synchronously. Would you expect that? I am not really sure what the behaviour should be like.

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js [hunk 6]
```diff
@@ -665,6 +665,12 @@ function updateOffscreenComponent(
   const prevState: OffscreenState | null =
     current !== null ? current.memoizedState : null;
 
+  // Offscreen stores pending changes to visibility in `_pendingVisibility`. This is
+  // to support batching of `attach` and `detach` calls.
+  workInProgress.stateNode._visibility &= ~OffscreenDetached;
+  workInProgress.stateNode._visibility |=
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberBeginWork.new.js:671]: This should happen in `commitWork`, not `beginWork`.

File: packages/react-reconciler/src/__tests__/ReactOffscreen-test.js (hunk 5 above)
Comment (acdlite) [reply packages/react-reconciler/src/__tests__/ReactOffscreen-test.js:1632]: The semantic model is that it behaves exactly as if you changed the `mode` prop with a state update (other than the fact that `detach` doesn't change any styles).

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js (hunk 6 above)
Comment (sammy-SC) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:671]: That wouldn't work, would it?

The visibility flag needs to be set before it is read for the first time. The visibility flag is read in begin phase, if it is set in commit work, that's too late. 
If we set it in commit phase, it would take effect on next update.

File: packages/react-reconciler/src/__tests__/ReactOffscreen-test.js (hunk 5 above)
Comment (sammy-SC) [reply packages/react-reconciler/src/__tests__/ReactOffscreen-test.js:1632]: I made sure attach/detach behaves the same as if `mode` prop was changed with a state update.

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js (hunk 6 above)
Comment (sammy-SC) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js]: Resolved in https://github.com/facebook/react/pull/25603/commits/bdf8abcbd8c59b610a5617f06dd12f55b0e509dc

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 7]
```diff
@@ -685,24 +685,18 @@ function updateOffscreenComponent(
 ) {
   const nextProps: OffscreenProps = workInProgress.pendingProps;
   const nextChildren = nextProps.children;
+  const isPendingDetached = (workInProgress.stateNode._pendingVisibility & OffscreenDetached) !== 0;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Nit: maybe `nextIsDetached` for consistency?

File: packages/react-reconciler/src/ReactFiberCommitWork.js [hunk 8]
```diff
@@ -2879,6 +2879,12 @@ function commitMutationEffectsOnFiber(
       // TODO: Add explicit effect flag to set _current.
       finishedWork.stateNode._current = finishedWork;
 
+      // Offscreen stores pending changes to visibility in `_pendingVisibility`. This is
+      // to support batching of `attach` and `detach` calls.
+      finishedWork.stateNode._visibility &= ~OffscreenDetached;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberCommitWork.js:2884]: Nit: assign the instance to a variable, since `stateNode` is any typed.

```js
const offscreenInstance: OffscreenInstance = finishedWork.stateNode;
```

File: packages/react-reconciler/src/ReactFiberBeginWork.js (hunk 7 above)
Comment (sammy-SC) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: Good idea!

File: packages/react-reconciler/src/ReactFiberCommitWork.js [hunk 9]
```diff
@@ -2857,12 +2876,19 @@ function commitMutationEffectsOnFiber(
       }
 
       commitReconciliationEffects(finishedWork);
+
+      const offscreenInstance: OffscreenInstance = finishedWork.stateNode;
+
       // TODO: Add explicit effect flag to set _current.
-      finishedWork.stateNode._current = finishedWork;
+      offscreenInstance._current = finishedWork;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberCommitWork.js:2883]: Let's make sure not to forget about this TODO (in a follow up)

File: packages/react-reconciler/src/ReactFiberCommitWork.js [hunk 10]
```diff
@@ -2407,24 +2408,42 @@ function getRetryCache(finishedWork) {
...
-  } else {
-    instance._visibility |= OffscreenDetached;
-    disappearLayoutEffects(currentOffscreenFiber);
-    disconnectPassiveEffect(currentOffscreenFiber);
+  if ((instance._pendingVisibility & OffscreenDetached) !== NoFlags) {
+    // The instance is already detached, this is a noop.
+    return;
+  }
+
+  const root = enqueueConcurrentRenderForLane(fiber, SyncLane);
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberCommitWork.js:2423]: Let's add a TODO to these methods that we can optimize in some cases by entering the commit phase directly without rendering.

**Review by acdlite — APPROVED** [review]
This looks good to me now but let's wait a day or two before landing in case there's an issue with the sync
